File: brady/happy_number.java
class Solution {
  
    // O(idk, i dont wanna go read number theory notes) time, O(once again, idk) space
    public boolean isHappy(int n) {
        Set<Integer> sums = new HashSet<>();
        while (true) {
            int sum = 0;
            while (n > 0) {
                int digit = n % 10;
                sum += digit * digit;
                n /= 10;
            }
            
            if (sum == 1) return true;
            if (sums.contains(sum)) return false;

            sums.add(sum);
            n = sum;
        }
    }
}

File: cole/day22_happy-number/happy-number.c
int ss(int n){
    int m = 0;
    while (n){
        m += (n % 10) * (n % 10);
        n /= 10;
    }
    return m;
}

bool isHappy(int n){
    int m = n;
    for (int i = 0; n > 1; i ^= 1) {
        n = ss(n);
        if (i)
            m = ss(m);
        if (n == m)
            return false;
    }
    return true;
}